### PR TITLE
Hunterization

### PR DESCRIPTION
**Pull request by FredrikAppelros**
Modify CMake config so that it can be used to produce a Hunter package.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -427,10 +420,8 @@ add_library (${LIB_NAME}
 target_compile_definitions(${LIB_NAME} PUBLIC "$<$<CONFIG:Debug>:ARABICA_DEBUG>")
 target_compile_definitions(${LIB_NAME} PUBLIC ARABICA_NOT_USE_PRAGMA_LINKER_OPTIONS)
 target_include_directories(${LIB_NAME} PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>)
-target_include_directories(${LIB_NAME} PUBLIC ${ADDITIONAL_INC} ${Boost_INCLUDE_DIRS})
 target_include_directories(${LIB_NAME} PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/include>)
-# link backend
-target_link_libraries(${LIB_NAME} ${ADDITIONAL_LIB})
+target_include_directories(${LIB_NAME} PUBLIC $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}>)
```
Comment (ruslo) [inline CMakeLists.txt]: Usually, it's just `include`, i.e. it should be:
```cmake
target_include_directories(${LIB_NAME} PUBLIC $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
```

Comment (ruslo) [reply CMakeLists.txt]: Also, you have `INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}` already in `install` command, so this one not needed.

File: CMakeLists.txt [hunk 2]
```diff
@@ -451,33 +442,73 @@ if(MSVC)
...
+  find_package(Boost CONFIG REQUIRED)
+  target_link_libraries(${LIB_NAME} Boost::boost)
+endif()
+
+#
+# Find Expat
+if(ARABICA_XML_BACKEND STREQUAL USE_EXPAT)
+  hunter_add_package(Expat)
+  find_package(EXPAT REQUIRED)
+  target_link_libraries(${LIB_NAME} ${EXPAT_LIBRARIES})
```
Comment (ruslo) [inline CMakeLists.txt]: There is an imported target:
- https://github.com/ruslo/hunter/blob/v0.23.155/examples/Expat/CMakeLists.txt#L18

We should use it. The documentation page seems to be out-of-date:
- https://docs.hunter.sh/en/latest/packages/pkg/Expat.html

Feel free to update :)

File: cmake/HunterGate.cmake [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+# This is a gate file to Hunter package manager.
```
Comment (ruslo) [inline cmake/HunterGate.cmake:25]: Please move this file into the separate commit, see "Perfect hunterization should ...":
- https://docs.hunter.sh/en/latest/creating-new/patch.html

**Review by ruslo — CHANGES_REQUESTED** [review]

File: cmake/Config.cmake.in [hunk 4]
```diff
@@ -0,0 +1,18 @@
+@PACKAGE_INIT@
+
+include(CMakeFindDependencyMacro)
+
+if(ARABICA_HAVE_BOOST)
+    find_dependency(Boost)
+endif()
+
+if(ARABICA_XML_BACKEND STREQUAL USE_EXPAT)
```
Comment (ruslo) [inline cmake/Config.cmake.in]: This file is created in a similar to [configure_file](https://cmake.org/cmake/help/latest/command/configure_file.html) way; it should be `@USE_EXPAT@`.

Comment (FredrikAppelros) [reply cmake/Config.cmake.in]: You mean `@ARABICA_XML_BACKEND@`, right?

Comment (ruslo) [reply cmake/Config.cmake.in]: I guess it should be both.

Comment (FredrikAppelros) [reply cmake/Config.cmake.in]: `USE_EXPAT` is a string which is one of the possible values of `ARABICA_XML_BACKEND`, hence it should not be expanded. I have changed it to use `@ARABICA_XML_BACKEND@` instead now but it still seems to fail with the same problem:
https://travis-ci.com/FredrikAppelros/hunter/jobs/194881832#L1147

File: cmake/Config.cmake.in [hunk 5]
```diff
@@ -0,0 +1,18 @@
+@PACKAGE_INIT@
+
+include(CMakeFindDependencyMacro)
+
+if(@ARABICA_HAVE_BOOST@)
+    find_dependency(Boost)
+endif()
+
+if(@ARABICA_XML_BACKEND@ STREQUAL USE_EXPAT)
```
Comment (ruslo) [inline cmake/Config.cmake.in]: It should be `@USE_EXPAT@`. Also, it will be more user-friendly to put everything in quotes. If both variables are empty, the generated line will be `if( STREQUAL )`. Not sure if it will work correctly, but it will be much easier to understand if you generate `if("" STREQUAL "")` instead.

File: cmake/Config.cmake.in [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    find_dependency(EXPAT)
+endif()
+
+if(@ARABICA_XML_BACKEND@ STREQUAL USE_LIBXML2)
```
Comment (ruslo) [inline cmake/Config.cmake.in]: `@USE_LIBXML2@`, see above.

**Review by ruslo — CHANGES_REQUESTED** [review]

File: cmake/Config.cmake.in (hunk 4 above)
Comment (ruslo) [reply cmake/Config.cmake.in]: We are generating `*Config.cmake` file for users. Variable `USE_EXPAT` from Arabica package build should be locked in `*Config.cmake`, hence it should be `@USE_EXPAT@`. Run the build and check the content of the `*Config.cmake` file.

Comment (FredrikAppelros) [reply cmake/Config.cmake.in]: As per my previous comment; `USE_EXPAT` is not a variable but a string which we compare the variable `ARABICA_XML_BACKEND` against.

File: cmake/Config.cmake.in (hunk 5 above)
Comment (FredrikAppelros) [reply cmake/Config.cmake.in]: Sure, I will wrap everything in quotes.

File: cmake/Config.cmake.in (hunk 4 above)
Comment (ruslo) [reply cmake/Config.cmake.in]: > As per my previous comment; USE_EXPAT is not a variable but a string

I got it now, what a poor choice of options... Definitely should be put in quotes then:
- https://cmake.org/cmake/help/latest/policy/CMP0054.html

Comment (FredrikAppelros) [reply cmake/Config.cmake.in]: I have wrapped everything in quotes but still no dice. The generated `arabicaConfig.cmake` file contains the following:
```
include(CMakeFindDependencyMacro)

if("")
    find_dependency(Boost)
endif()

if("USE_EXPAT" STREQUAL "USE_EXPAT")
    find_dependency(EXPAT)
endif()

if("USE_EXPAT" STREQUAL "USE_LIBXML2")
    find_dependency(libxml2)
endif()
```
Still the build fails with the same error:
> CMake Error at CMakeLists.txt:16 (add_executable):
  Target "boo" links to target "EXPAT::expat" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?

Any ideas what I am missing?

Comment (ruslo) [reply cmake/Config.cmake.in]: Probably `CONFIG`?
- https://github.com/ruslo/hunter/blob/v0.23.158/examples/Expat/CMakeLists.txt#L15

File: CMakeLists.txt [hunk 7]
```diff
@@ -451,33 +441,69 @@ if(MSVC)
...
 
 #
-# Add build dependency to boost if boost is part of the same project
-if(TARGET boost)
-  add_dependencies(${LIB_NAME} boost)
+# find libxml2
+if(ARABICA_XML_BACKEND STREQUAL USE_LIBXML2)
+  hunter_add_package(libxml2)
+  find_package(libxml2 CONFIG REQUIRED)
+  target_link_libraries(${LIB_NAME} ${LIBXML2_LIBRARIES})
```
Comment (ruslo) [inline CMakeLists.txt]: Please use the imported target:
- https://docs.hunter.sh/en/latest/packages/pkg/libxml2.html

File: CMakeLists.txt [hunk 8]
```diff
@@ -451,33 +441,69 @@ if(MSVC)
...
 # Install library
 #
-install(TARGETS ${LIB_NAME}
-  RUNTIME DESTINATION bin
-  LIBRARY DESTINATION lib
-  ARCHIVE DESTINATION lib/static
+install(
+  TARGETS ${LIB_NAME}
+  EXPORT ${TARGETS_EXPORT_NAME}
+  RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
```
Comment (ruslo) [inline CMakeLists.txt]: Basically, it's pretty the same. It is better to use `CMAKE_INSTALL_BINDIR` instead of `bin` in general but for the sake of minimal patch let's keep old values. Please send this change upstream first.

File: CMakeLists.txt [hunk 9]
```diff
@@ -451,33 +441,69 @@ if(MSVC)
...
+  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
 )
 
 install(DIRECTORY include/
-  DESTINATION include/${PROJECT_NAME}
+  DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}
```
Comment (ruslo) [inline CMakeLists.txt]: Same as above.

**Review by ruslo — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (FredrikAppelros) [reply CMakeLists.txt]: Unfortunately a lot of the headers internally refer to each other without using the `arabica` prefix. E.g. in `arabica/SAX/XMLReader.hpp` we can find the following:
```
#include <SAX/ArabicaConfig.hpp>
```

I have not been able to get this to work without some sort of equivalent to the above. But we can definitely make sure that we only use `INCLUDES DESTINATION include/${PROJECT_NAME}` instead of the construction seen above.

Comment (ruslo) [reply CMakeLists.txt]: > E.g. in arabica/SAX/XMLReader.hpp we can find the following

In my opinion, that's the wrong approach, but we can't do anything with it. It should be improved upstream.

> But we can definitely make sure that we only use INCLUDES DESTINATION include/${PROJECT_NAME}

Okay, that's fine in this case.

**Review by ruslo — APPROVED** [review]

**Comment by ruslo** [comment]
> Still the build fails with the same error:
> Any ideas what I am missing?

Does it work now, I can merge?

**Comment by FredrikAppelros** [comment]
Yes, now it works:
https://travis-ci.com/FredrikAppelros/hunter/builds/109404979
https://ci.appveyor.com/project/FredrikAppelros/hunter/builds/24064628

**Comment by ruslo** [comment]
Released: https://github.com/hunter-packages/arabica/releases/tag/v0.0.0-a202766-p0
